File: leetcode/DynamicProgram/leetcode-122/122.go
package dynamicprogram

func maxProfit(prices []int) int {
	if prices == nil || len(prices) < 2 {
		return 0
	}

	buy, profit := prices[0], 0
	for i := 1; i < len(prices); i++ {
		if prices[i] < prices[i-1] {
			profit += prices[i-1] - buy
			buy = prices[i]
		}
	}
	profit += prices[len(prices)-1] - buy
	return profit
}

File: leetcode/DynamicProgram/leetcode-122/122.java
class Solution {
    public int maxProfit(int[] prices) {
        if (prices == null || prices.length < 2) {
            return 0;
        }
        int buy = prices[0];
        int profit = 0;

        for (int i = 1; i < prices.length; i++) {
            if (prices[i] < prices[i - 1]) {
                profit += prices[i - 1] - buy;
                buy = prices[i];
            }
        }
        profit += prices[prices.length - 1] - buy;
        return profit;
    }
}